File: Course-Project/app/src/main/java/com/vladislavmyasnikov/courseproject/ui/main/AuthorizationActivity.java
package com.vladislavmyasnikov.courseproject.ui.main;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import com.google.android.material.textfield.TextInputEditText;
import com.vladislavmyasnikov.courseproject.R;
import com.vladislavmyasnikov.courseproject.data.models.Login;
import com.vladislavmyasnikov.courseproject.data.network.RequestResultListener;
import com.vladislavmyasnikov.courseproject.data.network.NetworkService;
import com.vladislavmyasnikov.courseproject.data.network.RequestResultCallback;

import androidx.appcompat.app.AppCompatActivity;
import okhttp3.Headers;
import retrofit2.Call;
import retrofit2.Response;

public class AuthorizationActivity extends AppCompatActivity implements RequestResultListener<Void> {

    public static final String COOKIES_STORAGE_NAME = "cookies_storage";
    public static final String AUTHORIZATION_TOKEN = "authorization_token";

    private TextInputEditText mEmailInputField;
    private TextInputEditText mPasswordInputField;
    private RequestResultCallback<Void> mRequestResultCallback = new RequestResultCallback<>(this);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_authorization);

        mEmailInputField = findViewById(R.id.input_email_field);
        mPasswordInputField = findViewById(R.id.input_password_field);
    }

    @Override
    public void onFailure(Call<Void> call, Throwable e) {
        Toast.makeText(this, R.string.no_internet_message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onResponse(Call<Void> call, Response<Void> response) {
        if (response.message().equals("OK")) {
            Headers headers = response.headers();

            String[] cookieData = headers.get("Set-Cookie").split("; ");
            String token = cookieData[0];

            SharedPreferences preferences = getSharedPreferences(COOKIES_STORAGE_NAME, Context.MODE_PRIVATE);
            preferences.edit()
                    .putString(AUTHORIZATION_TOKEN, token)
                    .apply();
            startWork();
        } else {
            Toast.makeText(this, R.string.incorrect_authorization_data_message, Toast.LENGTH_SHORT).show();
        }
    }

    private void startWork() {
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
        mRequestResultCallback.setRequestResultListener(null);
        finish();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mRequestResultCallback.setRequestResultListener(null);
    }

    public void onLoginClicked(View view) {
        String email = mEmailInputField.getText().toString();
        String password = mPasswordInputField.getText().toString();

        NetworkService.getInstance().getFintechService().getAccess(new Login(email, password)).enqueue(mRequestResultCallback);
    }
}

File: Course-Project/app/src/main/java/com/vladislavmyasnikov/courseproject/ui/main/AuthorizationActivity.kt
package com.vladislavmyasnikov.courseproject.ui.main

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProviders
import com.google.android.material.textfield.TextInputEditText
import com.vladislavmyasnikov.courseproject.R
import com.vladislavmyasnikov.courseproject.di.components.DaggerAuthorizationActivityInjector
import com.vladislavmyasnikov.courseproject.domain.repositories.*
import com.vladislavmyasnikov.courseproject.ui.viewmodels.LoginViewModel
import com.vladislavmyasnikov.courseproject.ui.viewmodels.LoginViewModelFactory
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Inject

class AuthorizationActivity : AppCompatActivity() {

    @Inject
    lateinit var viewModelFactory: LoginViewModelFactory

    private lateinit var loginViewModel: LoginViewModel
    private lateinit var emailInputField: TextInputEditText
    private lateinit var passwordInputField: TextInputEditText
    private lateinit var progressBar: ProgressBar
    private lateinit var loginButton: Button
    private val disposables = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_authorization)

        val component = DaggerAuthorizationActivityInjector.builder().appComponent(App.appComponent).build()
        component.injectAuthorizationActivity(this)
        loginViewModel = ViewModelProviders.of(this, viewModelFactory).get(LoginViewModel::class.java)

        emailInputField = findViewById(R.id.input_email_field)
        passwordInputField = findViewById(R.id.input_password_field)
        loginButton = findViewById(R.id.login_button)
        progressBar = findViewById(R.id.progress_bar)

        loginButton.setOnClickListener {
            val email = emailInputField.text.toString()
            val password = passwordInputField.text.toString()
            loginViewModel.login(email, password)
        }

        disposables.add(loginViewModel.loadingState.subscribe {
            setLoading(it)
        })

        disposables.add(loginViewModel.access.subscribe {
            startWork()
        })

        disposables.add(loginViewModel.errors.subscribe {
            when (it) {
                is IncorrectEmailInputException -> Toast.makeText(this, R.string.incorrect_input_email_message, Toast.LENGTH_SHORT).show()
                is IncorrectPasswordInputException -> Toast.makeText(this, R.string.incorrect_input_password_message, Toast.LENGTH_SHORT).show()
                is IncorrectLoginException -> Toast.makeText(this, R.string.incorrect_authorization_data_message, Toast.LENGTH_SHORT).show()
                is NoInternetException -> Toast.makeText(this, R.string.no_internet_message, Toast.LENGTH_SHORT).show()
            }
        })
    }

    override fun onDestroy() {
        super.onDestroy()
        disposables.clear()
    }

    private fun setLoading(isLoading: Boolean) {
        progressBar.visibility = if (isLoading) View.VISIBLE else View.INVISIBLE
        loginButton.isEnabled = !isLoading
    }

    private fun startWork() {
        val intent = Intent(this, MainActivity::class.java)
        startActivity(intent)
        finish()
    }
}
